crates/typst/src/foundations/args.rs

```diff
@@ -256,6 +256,56 @@ impl Args {
     }
 }
 
+/// A key that can be used to get an argument: either the index of a positional
+/// argument, or the name of a named argument.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum ArgumentKey {
+    Index(i64),
+    Name(Str),
+}
+
+cast! {
+    ArgumentKey,
+    v: i64 => Self::Index(v),
+    v: Str => Self::Name(v),
+}
+
+impl Args {
+    fn get(&self, key: &ArgumentKey) -> Option<Value> {
+        match key {
+            &ArgumentKey::Index(index) => {
+                if index < 0 {
+                    let index = (-(index + 1)).try_into().ok()?;
+                    self.items
+                        .iter()
+                        .filter(|item| item.name.is_none())
+                        .map(|item| item.value.v.clone())
+                        .rev()
+                        .nth(index)
+                } else {
+                    let index = index.try_into().ok()?;
+                    self.items
+                        .iter()
+                        .filter(|item| item.name.is_none())
+                        .map(|item| item.value.v.clone())
+                        .nth(index)
+                }
+            }
+
+            ArgumentKey::Name(name) => {
+                // Accept the last argument with the right name.
+                self.items
+                    .iter()
+                    .filter_map(|item| {
+                        item.name.as_ref().map(|k| (k, item.value.v.clone()))
+                    })
+                    .rfind(|(k, _)| *k == name)
+                    .map(|(_, v)| v)
+            }
+        }
+    }
```

Comment:
```suggestion
    fn get(&self, key: &ArgumentKey) -> Option<&Value> {
        let item = match key {
            &ArgumentKey::Index(index) => {
                let mut iter = self.items.iter().filter(|item| item.name.is_none());
                if index < 0 {
                    let index = (-(index + 1)).try_into().ok()?;
                    iter.nth_back(index)
                } else {
                    let index = index.try_into().ok()?;
                    iter.nth(index)
                }
            }

            // Accept the last argument with the right name.
            ArgumentKey::Name(name) => {
                self.items.iter().rfind(|item| item.name.as_ref() == Some(name))
            }
        };

        item.map(|item| &item.value.v)
    }
```
a bit less code duplication and potential cloning. requires `get(&key).cloned()` below

Reply:
This is indeed better.